File: src/recorder/src/computer-use/browser.js
// The browser harness: performs one step at a time in a page this recorder owns.
//
// The page lives in a BrowserWindow of its own (`browser-electron.js`), in a storage
// partition the employee's normal browsing never touches, and is driven over the Chrome
// DevTools Protocol. This module is Electron-free: it speaks to a `page` adapter
//
//   send(method, params)   CDP call
//   loadURL(url)           navigate and settle
//   url() / title()        what is loaded now
//   screenPoint(x, y)      page CSS pixel → screen pixel (for a real pointer), or null
//   close()
//
// and an optional `pointer` ({ moveTo(x, y), click() }) that moves the computer's own
// cursor; without one, clicks are dispatched to the page over CDP.
//
// What the server sees is what the accessibility tree says: a bounded list of named,
// interactive elements (`candidates`), the page's URL and title, and a text excerpt. A
// step may target only an element from the observation it cites; the harness keeps the
// mapping from candidate id to DOM node for the current observation only.
import { randomBytes } from 'node:crypto';

import { DEFAULT_SETTINGS } from '../recorder.js';
import { KEYS, isPrivateWindow } from './policy.js';

export const MAX_CANDIDATES = 40;
export const MAX_TEXT = 4000;
export const MAX_ROWS = 200;
export const MAX_COLUMNS = 40;
export const SETTLE_MS = 600;
export const HOVER_MS = 700; // real pointer rests on the control before pressing, so a watcher can follow

// Drawn into the sandbox page on every document: a ring that follows the pointer and
// flashes on press, so the person watching sees where the agent is and when it clicks.
// `aria-hidden` and `pointer-events:none` keep it out of the accessibility tree the
// harness enumerates and out of the way of the clicks themselves.
export const HALO_SCRIPT = `(() => {
  if (window.__vistaHalo) return;
  const halo = document.createElement('div');
  halo.id = 'vista-pointer-halo';
  halo.setAttribute('aria-hidden', 'true');
  halo.style.cssText = 'position:fixed;z-index:2147483647;pointer-events:none;width:56px;height:56px;margin:-28px 0 0 -28px;' +
    'border-radius:50%;border:4px solid #c2410c;background:rgba(194,65,12,.18);box-shadow:0 0 0 6px rgba(194,65,12,.25);' +
    'left:-100px;top:-100px;transition:transform .12s ease-out,background .12s;';
  const mount = () => (document.body || document.documentElement).appendChild(halo);
  if (document.body) mount(); else document.addEventListener('DOMContentLoaded', mount, { once: true });
  const at = (e) => { halo.style.left = e.clientX + 'px'; halo.style.top = e.clientY + 'px'; };
  window.addEventListener('mousemove', at, true);
  window.addEventListener('mousedown', (e) => {
    at(e);
    halo.style.transform = 'scale(.6)';
    halo.style.background = 'rgba(194,65,12,.6)';
    setTimeout(() => { halo.style.transform = ''; halo.style.background = 'rgba(194,65,12,.18)'; }, 220);
  }, true);
  window.__vistaHalo = true;
})();`;

// AX role → candidate kind. Anything not listed is not a target.
const ROLE_KIND = {
  textbox: 'field',
  searchbox: 'field',
  combobox: 'field',
  spinbutton: 'field',
  slider: 'field',
  link: 'link',
  button: 'interactive',
  checkbox: 'interactive',
  radio: 'interactive',
  switch: 'interactive',
  menuitem: 'interactive',
  menuitemcheckbox: 'interactive',
  menuitemradio: 'interactive',
  tab: 'interactive',
  option: 'interactive',
  treeitem: 'interactive',
  row: 'row',
  table: 'record',
  grid: 'record',
};

function axValue(prop) {
  const v = prop?.value;
  if (v == null) return null;
  return typeof v === 'object' ? v.value : v;
}

const CELL_ROLES = new Set(['cell', 'gridcell', 'rowheader']);
const ROW_NAME_CELLS = 3;

// Chrome gives a table row no accessible name; code derives one from its first cells so a
// row is a candidate the way a link or button is (its visible text, never anything hidden).
// Header rows (column headers only) are not offered.
function rowName(n, byId) {
  const cells = (n.childIds ?? []).map((id) => byId.get(id)).filter(Boolean);
  const data = cells.filter((c) => CELL_ROLES.has(axValue(c.role)));
  if (!data.length) return '';
  return data
    .map((c) => String(axValue(c.name) ?? '').trim())
    .filter(Boolean)
    .slice(0, ROW_NAME_CELLS)
    .join(' · ');
}

// Pure: AX nodes → candidates. Exported for tests.
export function candidatesFromAX(nodes, { limit = MAX_CANDIDATES } = {}) {
  const out = [];
  const seen = new Set();
  const byId = new Map((nodes ?? []).map((n) => [n.nodeId, n]));
  for (const n of nodes ?? []) {
    if (n.ignored || n.backendDOMNodeId == null) continue;
    const role = axValue(n.role);
    const kind = ROLE_KIND[role];
    if (!kind) continue;
    if (n.properties?.some((p) => p.name === 'hidden' && axValue(p) === true)) continue;
    const name = String(axValue(n.name) || (role === 'row' ? rowName(n, byId) : ''))
      .replace(/\s+/g, ' ')
      .trim()
      .slice(0, 200);
    if (!name) continue;
    const key = `${role}\0${name.toLowerCase()}`;
    if (seen.has(key)) continue;
    seen.add(key);
    const attrs = {};
    const disabled = n.properties?.find((p) => p.name === 'disabled');
    if (disabled && axValue(disabled) === true) attrs.disabled = true;
    const val = axValue(n.value);
    if (kind === 'field' && val != null && String(val).length) attrs.has_value = true; // never the value itself
    out.push({ id: String(n.backendDOMNodeId), role, name, kind, attrs });
    if (out.length >= limit) break;
  }
  return out;
}

export function observationId() {
  return `obs-${randomBytes(6).toString('hex')}`;
}

export class BrowserHarness {
  // `open` creates the page adapter on first use (so no window exists until a session
  // needs one); `pointer` is optional; `settings` supplies the private app/title lists.
  constructor({ open, pointer = null, settings = () => DEFAULT_SETTINGS, sleep = (ms) => new Promise((r) => setTimeout(r, ms)) }) {
    this.kind = 'browser';
    this.supported = typeof open === 'function';
    this.open = open;
    this.pointer = pointer;
    this.settings = settings;
    this.sleep = sleep;
    this.page = null;
    this.current = null; // { id, nodes: Map<candidate id, backendNodeId> }
  }

  capabilities() {
    return this.supported ? ['observe', 'navigate', 'click', 'type', 'press', 'extract', 'screenshot', 'wait'] : [];
  }

  async perform(step) {
    try {
      const page = await this._page();
      switch (step.action) {
        case 'observe':
          return this._ok(step, 'Looked at the page', { observation: await this._observe(page) });
        case 'navigate': {
          await page.loadURL(step.value);
          await this.sleep(SETTLE_MS);
          const observation = await this._observe(page);
          return this._ok(step, `Opened ${observation.sensitive ? 'a page' : observation.title || step.value}`, { observation, result: { url_after: observation.url, title_after: observation.title } });
        }
        case 'click': {
          const node = this._node(step);
          const box = await this._center(page, node);
          await this._click(page, box);
          await this.sleep(SETTLE_MS);
          const observation = await this._observe(page);
          return this._ok(step, step.description || 'Clicked the control', { observation, result: { url_after: observation.url, title_after: observation.title } });
        }
        case 'type': {
          const node = this._node(step);
          await this._click(page, await this._center(page, node));
          const previous = await this._fill(page, node, step.value, step.replace !== false);
          const observation = await this._observe(page);
          return this._ok(step, step.description || 'Typed the value into the field', {
            observation,
            result: { previous_value: previous, undo: { action: 'type', target_id: step.target_id, value: previous ?? '' } },
          });
        }
        case 'press': {
          await this._press(page, step.value);
          await this.sleep(SETTLE_MS);
          const observation = await this._observe(page);
          return this._ok(step, `Pressed ${step.value}`, { observation, result: { url_after: observation.url, title_after: observation.title } });
        }
        case 'extract': {
          await this._guard(page);
          const node = step.target_id != null ? this._node(step) : null;
          const result = step.value === 'table' ? await this._table(page, node) : { text: await this._text(page, node) };
          return this._ok(step, step.value === 'table' ? `Extracted ${result.count} rows` : 'Read the text', { observation: await this._observe(page), result });
        }
        case 'screenshot': {
          await this._guard(page);
          const shot = await page.send('Page.captureScreenshot', { format: 'png' });
          return this._ok(step, 'Took a screenshot', { observation: await this._observe(page), evidence: { screenshot_base64: shot.data, content_type: 'image/png' } });
        }
        case 'wait':
          await this.sleep(Number(step.value));
          return this._ok(step, `Waited ${Number(step.value)} ms`, { observation: await this._observe(page) });
        default:
          return this._fail(step, 'invalid_value', `The browser harness cannot ${step.action}.`);
      }
    } catch (e) {
      return this._fail(step, e.code ?? 'harness_error', e.message);
    }
  }

  async close() {
    const p = this.page;
    this.page = null;
    this.current = null;
    if (p) await p.close();
  }

  // ---- observing ----

  async _observe(page) {
    const url = await page.url();
    const title = await page.title();
    const id = observationId();
    if (isPrivateWindow('browser', title, this.settings()) || isPrivateWindow('browser', url, this.settings())) {
      this.current = { id, nodes: new Map() };
      return { observation_id: id, url: '', title: '', sensitive: true, candidates: [], text_excerpt: '' };
    }
    const { nodes } = await page.send('Accessibility.getFullAXTree', {});
    const candidates = candidatesFromAX(nodes);
    this.current = { id, nodes: new Map(candidates.map((c) => [c.id, Number(c.id)])) };
    return { observation_id: id, url, title, sensitive: false, candidates, text_excerpt: await this._text(page, null) };
  }

  async _guard(page) {
    const title = await page.title();
    const url = await page.url();
    if (isPrivateWindow('browser', title, this.settings()) || isPrivateWindow('browser', url, this.settings())) {
      const err = new Error('This page looks private (sign-in, payment or a listed app); the step was not performed.');
      err.code = 'sensitive_window';
      throw err;
    }
  }

  _node(step) {
    const backendNodeId = this.current?.id === step.observation_id ? this.current.nodes.get(String(step.target_id)) : undefined;
    if (backendNodeId == null) {
      const err = new Error('The target is not in the current observation.');
      err.code = 'stale_observation';
      throw err;
    }
    return backendNodeId;
  }

  // ---- acting ----

  async _center(page, backendNodeId) {
    await this._guard(page);
    await page.send('DOM.scrollIntoViewIfNeeded', { backendNodeId }).catch(() => {});
    const { model } = await page.send('DOM.getBoxModel', { backendNodeId });
    const q = model.content; // [x1,y1, x2,y2, x3,y3, x4,y4]
    return { x: (q[0] + q[4]) / 2, y: (q[1] + q[5]) / 2 };
  }

  async _click(page, { x, y }) {
    await page.send('Runtime.evaluate', { expression: HALO_SCRIPT }).catch(() => {});
    const screen = this.pointer && page.screenPoint ? await page.screenPoint(x, y) : null;
    if (screen) {
      await this.pointer.moveTo(screen.x, screen.y);
      await this.sleep(HOVER_MS);
      await this.pointer.click();
      return;
    }
    await page.send('Input.dispatchMouseEvent', { type: 'mouseMoved', x, y });
    await page.send('Input.dispatchMouseEvent', { type: 'mousePressed', x, y, button: 'left', clickCount: 1 });
    await page.send('Input.dispatchMouseEvent', { type: 'mouseReleased', x, y, button: 'left', clickCount: 1 });
  }

  async _fill(page, backendNodeId, text, replace) {
    await this._guard(page);
    const { object } = await page.send('DOM.resolveNode', { backendNodeId });
    const read = await page.send('Runtime.callFunctionOn', {
      objectId: object.objectId,
      functionDeclaration: 'function(){ this.focus(); return this.isContentEditable ? this.textContent : (this.value ?? null); }',
      returnByValue: true,
    });
    const previous = read.result?.value ?? null;
    if (replace) {
      await page.send('Runtime.callFunctionOn', {
        objectId: object.objectId,
        functionDeclaration: 'function(){ if (this.select) this.select(); else { const r = document.createRange(); r.selectNodeContents(this); const s = getSelection(); s.removeAllRanges(); s.addRange(r); } }',
      });
    }
    await page.send('Input.insertText', { text });
    return previous;
  }

  async _press(page, name) {
    await this._guard(page);
    const k = KEYS[name];
    const base = { key: k.key, code: k.code, windowsVirtualKeyCode: k.keyCode, nativeVirtualKeyCode: k.keyCode };
    await page.send('Input.dispatchKeyEvent', { type: 'keyDown', ...base });
    await page.send('Input.dispatchKeyEvent', { type: 'keyUp', ...base });
  }

  async _text(page, backendNodeId) {
    if (backendNodeId != null) {
      const { object } = await page.send('DOM.resolveNode', { backendNodeId });
      const r = await page.send('Runtime.callFunctionOn', { objectId: object.objectId, functionDeclaration: 'function(){ return (this.innerText ?? this.textContent ?? "") }', returnByValue: true });
      return String(r.result?.value ?? '').slice(0, MAX_TEXT);
    }
    const expression = `(document.body && document.body.innerText || '').slice(0, ${MAX_TEXT})`;
    const r = await page.send('Runtime.evaluate', { expression, returnByValue: true });
    return String(r.result?.value ?? '').slice(0, MAX_TEXT);
  }

  async _table(page, backendNodeId) {
    const fn = `function(){
      const root = this instanceof Element ? this : document;
      const table = root.tagName === 'TABLE' ? root : root.querySelector('table');
      if (!table) return { columns: [], rows: [], count: 0 };
      const cell = (c) => (c.innerText || c.textContent || '').trim().slice(0, 200);
      const trs = Array.from(table.querySelectorAll('tr'));
      let columns = Array.from(trs[0]?.querySelectorAll('th') ?? []).map(cell).slice(0, ${MAX_COLUMNS});
      let body = columns.length ? trs.slice(1) : trs;
      if (!columns.length) columns = Array.from(trs[0]?.children ?? []).map((_, i) => 'column_' + (i + 1)).slice(0, ${MAX_COLUMNS});
      const rows = body.slice(0, ${MAX_ROWS}).map((tr) => Array.from(tr.children).slice(0, ${MAX_COLUMNS}).map(cell));
      return { columns, rows, count: rows.length };
    }`;
    if (backendNodeId != null) {
      const { object } = await page.send('DOM.resolveNode', { backendNodeId });
      const r = await page.send('Runtime.callFunctionOn', { objectId: object.objectId, functionDeclaration: fn, returnByValue: true });
      return r.result.value;
    }
    const r = await page.send('Runtime.evaluate', { expression: `(${fn}).call(null)`, returnByValue: true });
    return r.result.value;
  }

  // ---- plumbing ----

  async _page() {
    if (!this.supported) {
      const err = new Error('The browser harness is not available on this computer.');
      err.code = 'harness_unsupported';
      throw err;
    }
    if (!this.page) {
      this.page = await this.open();
      await this.page.send('Page.addScriptToEvaluateOnNewDocument', { source: HALO_SCRIPT }).catch(() => {});
    }
    return this.page;
  }

  _ok(step, description, { observation = null, result = null, evidence = null } = {}) {
    return { ok: true, description: description || step.description || step.action, observation, result, evidence, error: null };
  }

  _fail(step, code, message) {
    return { ok: false, description: `${step.action} refused: ${message}`, observation: null, result: null, evidence: null, error: { code, message } };
  }
}

File: src/vista_device/drivers/browser.py
"""Browser driver on browser-use's `BrowserSession` (observation + action layers only).

What is taken from browser-use: the merged DOM/AX snapshot (`get_browser_state_summary` →
`selector_map` of `EnhancedDOMTreeNode`), its event bus for clicks/typing/keys/navigation,
`allowed_domains` enforcement and the isolated profile. What is *not* taken: `Agent`, the
system prompt, any LLM client, telemetry. The library is pinned in `pyproject.toml`
(`[device]` extra) and is imported lazily so a recorder build without it still runs the
frame/leakage code paths.

`_candidates()` is the pure part (a list of node views → `Candidate`s) and is what the tests
cover; the session plumbing is exercised by the recorded device self-test only.
"""

from __future__ import annotations

import asyncio
import re
from collections.abc import Iterable
from dataclasses import dataclass

from taskmining.normalise import Vocabulary
from vista_device.drivers import ACTIONS, DriverError, Result
from vista_device.frame import DIALOG_ROLES, LANDMARK_ROLES, MAX_TEXT, Candidate, Frame, frame_from_candidates, position_class
from vista_device.settle import settle

try:
    from browser_use import BrowserSession
    from browser_use.browser.events import ClickElementEvent, NavigateToUrlEvent, SendKeysEvent, TypeTextEvent
except ImportError:  # the [device] extra is optional
    BrowserSession = None
    ClickElementEvent = NavigateToUrlEvent = SendKeysEvent = TypeTextEvent = None

MAX_CANDIDATES = 40
ROLE_KIND: dict[str, str] = {
    "textbox": "field",
    "searchbox": "field",
    "combobox": "field",
    "spinbutton": "field",
    "slider": "field",
    "link": "link",
    "button": "interactive",
    "checkbox": "interactive",
    "radio": "interactive",
    "switch": "interactive",
    "menuitem": "interactive",
    "menuitemcheckbox": "interactive",
    "menuitemradio": "interactive",
    "tab": "interactive",
    "option": "interactive",
    "treeitem": "interactive",
    "row": "row",
    "table": "record",
    "grid": "record",
}
TAG_ROLE: dict[str, str] = {
    "a": "link",
    "button": "button",
    "input": "textbox",
    "textarea": "textbox",
    "select": "combobox",
    "tr": "row",
    "table": "table",
    "nav": "navigation",
    "main": "main",
    "header": "banner",
    "footer": "contentinfo",
    "form": "form",
    "dialog": "dialog",
    "aside": "complementary",
}
INPUT_TYPE_ROLE: dict[str, str] = {
    "checkbox": "checkbox",
    "radio": "radio",
    "button": "button",
    "submit": "button",
    "search": "searchbox",
    "range": "slider",
    "number": "spinbutton",
}
KEY_NAMES: dict[str, str] = {
    "Enter": "Enter",
    "Tab": "Tab",
    "Escape": "Escape",
    "ArrowDown": "ArrowDown",
    "ArrowUp": "ArrowUp",
    "Backspace": "Backspace",
    "Space": "Space",
}
_SPACE = re.compile(r"\s+")


@dataclass
class NodeView:
    """The slice of an `EnhancedDOMTreeNode` the adapter reads; built by `node_view()` from
    the library object, or directly by tests."""

    index: int
    backend_node_id: int
    tag: str
    role: str | None
    name: str
    attributes: dict[str, str]
    ancestor_roles: tuple[str, ...]
    x: float | None = None
    y: float | None = None
    visible: bool = True
    disabled: bool = False
    has_value: bool = False


def role_of(tag: str, ax_role: str | None, attributes: dict[str, str]) -> str:
    explicit = (attributes.get("role") or "").strip().lower()
    if ax_role and ax_role not in {"generic", "none", "presentation", ""}:
        return ax_role
    if explicit:
        return explicit
    if tag == "input":
        return INPUT_TYPE_ROLE.get((attributes.get("type") or "text").lower(), "textbox")
    return TAG_ROLE.get(tag, tag)


def node_view(index: int, node: object) -> NodeView:  # node: browser_use EnhancedDOMTreeNode
    ax = node.ax_node
    attributes = dict(node.attributes or {})
    tag = node.tag_name.lower()
    name = (
        (ax.name if ax and ax.name else "")
        or attributes.get("aria-label")
        or attributes.get("placeholder")
        or attributes.get("title")
        or ""
    )
    if not name and tag not in {"input", "textarea", "select"}:
        name = node.get_all_children_text(max_depth=2)
    ancestors: list[str] = []
    p = node.parent_node
    while p is not None:
        pax = p.ax_node
        r = role_of(p.tag_name.lower(), pax.role if pax else None, dict(p.attributes or {}))
        if r in LANDMARK_ROLES or r in DIALOG_ROLES:
            ancestors.append(r)
        p = p.parent_node
    pos = node.absolute_position
    props = {pr.name: pr.value for pr in (ax.properties or [])} if ax else {}
    has_value = bool(attributes.get("value")) or bool(props.get("valuetext")) or (props.get("checked") is True)
    return NodeView(
        index=index,
        backend_node_id=node.backend_node_id,
        tag=tag,
        role=ax.role if ax else None,
        name=name,
        attributes=attributes,
        ancestor_roles=tuple(ancestors),
        x=pos.x if pos else None,
        y=pos.y if pos else None,
        visible=node.is_visible is not False,
        disabled=props.get("disabled") is True or "disabled" in attributes,
        has_value=has_value,
    )


def candidates_from_views(
    views: Iterable[NodeView], *, viewport: tuple[int, int] | None, limit: int = MAX_CANDIDATES
) -> tuple[list[Candidate], list[str], str | None]:
    """Views → (candidates, landmark roles present, front dialog name). Pure."""
    out: list[Candidate] = []
    seen: set[str] = set()
    landmarks: set[str] = set()
    dialog: str | None = None
    vw, vh = viewport or (None, None)
    for v in views:
        role = role_of(v.tag, v.role, v.attributes)
        landmarks.update(a for a in v.ancestor_roles if a in LANDMARK_ROLES)
        if role in LANDMARK_ROLES:
            landmarks.add(role)
        if role in DIALOG_ROLES and dialog is None:
            dialog = v.name or ""
        if any(a in DIALOG_ROLES for a in v.ancestor_roles) and dialog is None:
            dialog = ""
        kind = ROLE_KIND.get(role)
        if not kind or not v.visible:
            continue
        name = _SPACE.sub(" ", v.name or "").strip()[:200]
        if not name:
            continue
        key = f"{role}\0{name.lower()}"
        if key in seen:
            continue
        seen.add(key)
        landmark = next((a for a in v.ancestor_roles if a in LANDMARK_ROLES), "")
        out.append(
            Candidate(
                id=str(v.index),
                role=role,
                name=name,
                kind=kind,
                landmark=landmark,
                position=position_class(v.x, v.y, vw, vh),
                has_value=v.has_value if kind == "field" else False,
                disabled=v.disabled,
                primary=(v.attributes.get("type") or "").lower() == "submit",
            )
        )
        if len(out) >= limit:
            break
    return out, sorted(landmarks), dialog


class BrowserDriver:
    kind = "browser"

    def __init__(
        self,
        *,
        allowed_domains: list[str] | None = None,
        user_data_dir: str | None = None,
        headless: bool = False,
        vocab: Vocabulary | None = None,
        screenshots: bool = False,
    ):
        if BrowserSession is None:
            raise DriverError("harness_unsupported", "browser-use is not installed in this recorder build (pip install 'vista[device]').")
        self.session = BrowserSession(allowed_domains=allowed_domains, user_data_dir=user_data_dir, headless=headless)
        self.vocab = vocab
        self.screenshots = screenshots
        self._started = False
        self._nodes: dict[str, object] = {}
        self._frame: Frame | None = None

    def capabilities(self) -> list[str]:
        return list(ACTIONS)

    async def _start(self) -> None:
        if not self._started:
            await self.session.start()
            self._started = True

    async def _fingerprint(self) -> str:
        state = await self.session.get_browser_state_summary(include_screenshot=False)
        return "|".join(
            f"{i}:{n.tag_name}:{(n.ax_node.name if n.ax_node else '') or ''}" for i, n in sorted(state.dom_state.selector_map.items())
        )

    async def observe(self) -> Frame:
        await self._start()
        settled = await settle(self._fingerprint)
        state = await self.session.get_browser_state_summary(include_screenshot=False)
        views = [node_view(i, n) for i, n in sorted(state.dom_state.selector_map.items())]
        viewport = (state.page_info.viewport_width, state.page_info.viewport_height) if state.page_info else None
        candidates, landmarks, dialog = candidates_from_views(views, viewport=viewport)
        self._nodes = {str(i): n for i, n in state.dom_state.selector_map.items()}
        text = ""
        page = await self.session.get_current_page()
        if page is not None:
            text = str(await page.evaluate("() => (document.body && document.body.innerText) || ''"))[:MAX_TEXT]
        self._frame = frame_from_candidates(
            kind="browser",
            url=state.url,
            title=state.title,
            candidates=candidates,
            landmarks=landmarks,
            dialog=dialog,
            text=text,
            vocab=self.vocab,
            settled=settled,
        )
        return self._frame

    def _node(self, step: dict) -> object:
        node = self._nodes.get(str(step.get("target_id")))
        if node is None:
            raise DriverError("stale_observation", "The target is not in the current observation.")
        return node

    async def perform(self, step: dict, frame: Frame | None) -> Result:
        action = str(step.get("action", ""))
        try:
            await self._start()
            bus = self.session.event_bus
            if action == "observe":
                return Result(True, "Looked at the page", frame=await self.observe())
            if action == "navigate":
                await bus.dispatch(NavigateToUrlEvent(url=str(step["value"])))
                f = await self.observe()
                return Result(
                    True,
                    "Opened a page" if f.sensitive else f"Opened {f.title or step['value']}",
                    frame=f,
                    result={"url_after": f.url, "title_after": f.title},
                )
            if action == "click":
                await bus.dispatch(ClickElementEvent(node=self._node(step)))
                f = await self.observe()
                return Result(
                    True, step.get("description") or "Clicked the control", frame=f, result={"url_after": f.url, "title_after": f.title}
                )
            if action == "type":
                node = self._node(step)
                previous = str(dict(node.attributes or {}).get("value") or "")
                await bus.dispatch(
                    TypeTextEvent(node=node, text=str(step["value"]), clear=step.get("replace", True) is not False, is_sensitive=True)
                )
                f = await self.observe()
                return Result(
                    True,
                    step.get("description") or "Typed the value into the field",
                    frame=f,
                    result={"previous_value": previous, "undo": {"action": "type", "target_id": step.get("target_id"), "value": previous}},
                )
            if action == "press":
                key = KEY_NAMES.get(str(step.get("value")))
                if key is None:
                    return Result.refused(action, "invalid_value", f"Unknown key {step.get('value')!r}.")
                await bus.dispatch(SendKeysEvent(keys=key))
                f = await self.observe()
                return Result(True, f"Pressed {key}", frame=f, result={"url_after": f.url, "title_after": f.title})
            if action == "extract":
                f = await self.observe()
                return Result(True, "Read the text", frame=f, result={"text": f.text_local})
            if action == "screenshot":
                if not self.screenshots:
                    return Result.refused(action, "consent_required", "Screenshots are not covered by this session's consent.")
                page = await self.session.must_get_current_page()
                data = await page.screenshot(format="png")
                return Result(
                    True, "Took a screenshot", frame=await self.observe(), evidence={"screenshot_base64": data, "content_type": "image/png"}
                )
            if action == "wait":
                await asyncio.sleep(float(step.get("value", 0)) / 1000)
                return Result(True, f"Waited {step.get('value')} ms", frame=await self.observe())
            return Result.refused(action, "invalid_value", f"The browser driver cannot {action}.")
        except DriverError as e:
            return Result.refused(action, e.code, str(e))
        except Exception as e:  # noqa: BLE001 - the run loop needs a refusal, not a crash
            return Result.refused(action, "harness_error", str(e)[:500])

    async def close(self) -> None:
        if self._started:
            await self.session.stop()
            self._started = False
